[BUG] Spark 3.1.1 test failure in writing due to removal of InMemoryFileIndex.shouldFilterOut

**Describe the bug**
[SPARK-34075][SQL][CORE] Hidden directories are being listed for partiition inference
removed shouldFilterOut from InMemoryFileIndex and now the orc and parquet write tests are failing:

06:25:49  FAILED src/main/python/orc_write_test.py::test_write_save_table[native-[Byte, Short, Integer, Long, Float, Double, String, Boolean, Date, Timestamp]]
06:25:49  FAILED src/main/python/orc_write_test.py::test_write_save_table[hive-[Byte, Short, Integer, Long, Float, Double, String, Boolean, Date, Timestamp]]
06:25:49  FAILED src/main/python/orc_write_test.py::test_write_sql_save_table[native-TIMESTAMP_MICROS-[Byte, Short, Integer, Long, Float, Double, String, Boolean, Date, Timestamp]]
06:25:49  FAILED src/main/python/orc_write_test.py::test_write_sql_save_table[native-TIMESTAMP_MILLIS-[Byte, Short, Integer, Long, Float, Double, String, Boolean, Date, Timestamp]]
06:25:49  FAILED src/main/python/orc_write_test.py::test_write_sql_save_table[hive-TIMESTAMP_MICROS-[Byte, Short, Integer, Long, Float, Double, String, Boolean, Date, Timestamp]]
06:25:49  FAILED src/main/python/orc_write_test.py::test_write_sql_save_table[hive-TIMESTAMP_MILLIS-[Byte, Short, Integer, Long, Float, Double, String, Boolean, Date, Timestamp]]
06:25:49  FAILED src/main/python/parquet_write_test.py::test_write_save_table[INT96-[Byte, Short, Integer, Long, Float, Double, String, Boolean, Date, Timestamp]]
06:25:49  FAILED src/main/python/parquet_write_test.py::test_write_save_table[TIMESTAMP_MICROS-[Byte, Short, Integer, Long, Float, Double, String, Boolean, Date, Timestamp]]
06:25:49  FAILED src/main/python/parquet_write_test.py::test_write_save_table[TIMESTAMP_MILLIS-[Byte, Short, Integer, Long, Float, Double, String, Boolean, Date, Timestamp]]
06:25:49  FAILED src/main/python/parquet_write_test.py::test_write_sql_save_table[INT96-[Byte, Short, Integer, Long, Float, Double, String, Boolean, Date, Timestamp]]
06:25:49  FAILED src/main/python/parquet_write_test.py::test_write_sql_save_table[TIMESTAMP_MICROS-[Byte, Short, Integer, Long, Float, Double, String, Boolean, Date, Timestamp]]
06:25:49  FAILED src/main/python/parquet_write_test.py::test_write_sql_save_table[TIMESTAMP_MILLIS-[Byte, Short, Integer, Long, Float, Double, String, Boolean, Date, Timestamp]]
06:25:49  FAILED src/main/python/parquet_write_test.py::test_ts_write_twice_fails_exception

--- a/shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/Spark300Shims.scala
+++ b/shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/Spark300Shims.scala
@@ -35,7 +35,7 @@ import org.apache.spark.sql.catalyst.rules.Rule
 import org.apache.spark.sql.connector.read.Scan
 import org.apache.spark.sql.execution._
 import org.apache.spark.sql.execution.adaptive.ShuffleQueryStageExec
-import org.apache.spark.sql.execution.datasources.{FilePartition, FileScanRDD, HadoopFsRelation, PartitionDirectory, PartitionedFile}
+import org.apache.spark.sql.execution.datasources.{FilePartition, FileScanRDD, HadoopFsRelation, InMemoryFileIndex, PartitionDirectory, PartitionedFile}
 import org.apache.spark.sql.execution.datasources.v2.orc.OrcScan
 import org.apache.spark.sql.execution.datasources.v2.parquet.ParquetScan
 import org.apache.spark.sql.execution.exchange.{BroadcastExchangeExec, ShuffleExchangeExec}
@@ -435,5 +435,9 @@ class Spark300Shims extends SparkShims {
       qualifier: Seq[String],
       explicitMetadata: Option[Metadata]): Alias = {
     Alias(child, name)(exprId, qualifier, explicitMetadata)
+  }
+
+  override def shouldIgnorePath(path: String): Boolean = {
+    InMemoryFileIndex.shouldFilterOut(path)
   }
 }

--- a/shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/Spark311Shims.scala
+++ b/shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/Spark311Shims.scala
@@ -376,4 +376,8 @@ class Spark311Shims extends Spark301Shims {
       explicitMetadata: Option[Metadata]): Alias = {
     Alias(child, name)(exprId, qualifier, explicitMetadata)
   }
+
+  override def shouldIgnorePath(path: String): Boolean = {
+    HadoopFSUtilsShim.shouldIgnorePath(path)
+  }
 }

--- a/shims/spark311/src/main/scala/org/apache/spark/sql/rapids/shims/spark311/HadoopFSUtilsShim.scala
+++ b/shims/spark311/src/main/scala/org/apache/spark/sql/rapids/shims/spark311/HadoopFSUtilsShim.scala
@@ -1,0 +1,25 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids.shims.spark311
+
+import org.apache.spark.util.HadoopFSUtils
+
+object HadoopFSUtilsShim {
+
+  def shouldIgnorePath(path: String) = HadoopFSUtils.shouldFilterOutPathName(path)
+
+}

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/SparkShims.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/SparkShims.scala
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, NVIDIA CORPORATION.
+ * Copyright (c) 2020-2021, NVIDIA CORPORATION.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/SparkShims.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/SparkShims.scala
@@ -163,4 +163,6 @@ trait SparkShims {
       exprId: ExprId,
       qualifier: Seq[String] = Seq.empty,
       explicitMetadata: Option[Metadata] = None): Alias
+
+  def shouldIgnorePath(path: String): Boolean
 }

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuDataSource.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuDataSource.scala
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, NVIDIA CORPORATION.
+ * Copyright (c) 2020-2021, NVIDIA CORPORATION.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuDataSource.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuDataSource.scala
@@ -638,7 +638,7 @@ object GpuDataSource extends Logging {
     val allPaths = globbedPaths ++ nonGlobPaths
     if (checkFilesExist) {
       val (filteredOut, filteredIn) = allPaths.partition { path =>
-        InMemoryFileIndex.shouldFilterOut(path.getName)
+        ShimLoader.getSparkShims.shouldIgnorePath(path.getName)
       }
       if (filteredIn.isEmpty) {
         logWarning(
@@ -649,7 +649,7 @@ object GpuDataSource extends Logging {
       }
     }
 
-    allPaths.toSeq
+    allPaths
   }
 
 }

